leo-editor/leo-editor
# PR: Rewrite Importer.gen_blocks and helpers

Félix: The `var_equal_function` unit test was ignored because it didn't start with `test_`!

This PR rewrites  `i.gen_blocks` as an iterative pipeline. The recursive version was fatally flawed!

The generated code has changed. Unit tests have changed accordingly.
- The code generators no longer generate section references for docstrings or declarations.
  As before, the code generators move module-level and class docstrings to their proper places.
- All nodes start with no leading newlines and end with exactly one newline.
- Several importers generate improved headlines in special situations.

This PR adds one method and one function. They should have been added years ago!
- Add `p.insertAsFirstChild` to match existing `v.insertAsFirstChild`.
- Add `g.my_name`. Example: `g.printObj(o, tag=f"{my_name()}: {p.h}")`.

<details><summary>Code Changes</summary>
<br>

- [x] Make `i.gen_block` an iterative pipeline using `i.generate_all_blocks` and `python_i.postprocess`.
- [x] Use a `Block` *class* instead of a tuple/namedtuple.
- [x] Remove the `Block` argument from `gen_block` methods in all importers.
- [x] Remove the `allow_preamble` ivar. `i.generate_all_blocks` always calls `x.postprocess`.
- [x] `i.compute_body` removes leading and trailing blank lines from *all* generated bodies.
- [x] `python_i.postprocess` uses helpers *functions* instead of helper methods.
- [x] Improve the unit-testing infrastructure in minor ways.
- [x] The Python importer generates `function:` in headlines for non-method defs.
- [x] Add `p.insertAsFirstChild` to match existing `v.insertAsFirstChild`.
- [x] Add `g.my_name`.

</details>

<details><summary>Testing</summary>
<br>

Many unit tests changed, most in minor ways. See the diffs for full details.

`i.compute_body` removes leading and trailing blank lines from *all* generated bodies.
*Unit tests have changed accordingly.*

The new importers pass the following **round-trip test** on mypy's sources:

- Import all of mypy's sources using the recursive import script.
- Write all files with `write-at-file-nodes` (`Ctrl-Shift-W').
- Close Leo.
- Revert all files with `git checkout .`.
- Reopen Leo and redo `write-at-file-nodes`.
- Observe that no files change.
- Close Leo
- Observer that `git status` reports no files have changed.

**New and deleted tests**

- Add `TestPython.test_post_process_long_outer_docstring`.
- Delete `TestHtml.test_slideshow_slide`. This test was an impossible-to-understand hack.
- Delete `TestPython.test_preamble`. Leo no longer generates section references for docstrings or declarations.

</details>
